_posts/2019-03-28-smart-proxy-capabilities-explained.md

```diff
@@ -0,0 +1,205 @@
+
+While this part wasn't part of my initial design, the new datastructure easily allowed this almost for free.
+
+### Showing capabilities
+
+This was the original reason for the whole change, which is why I still think of this as the capabilities effort. Interestingly enough the other fields have seen more use. Exposing failed modules has allowed a colleague to write a script assisting in debugging installations. Exposing settings has seen adoption in a few places. This is because they're very easy to use. Capabilities require much more support on the Foreman side.
+
+My initial desire was to expand the DNS implementation. Some providers are capable of managing zones themselves. PowerDNS and Infoblox provide full APIs for this and these could be used. However, in the classic scenario there was no way of knowing whether the Proxy actually supported this other than trying. This is not a good user experience. If the Proxy could tell if it had the `DNS_ZONE_MANAGEMENT` capability, we could tell the user the domain was actually created. If it didn't we could warn the user they needed to manually ensure the domain was created. Perhaps more fine grained access is also possible, but the idea is still that simple strings indicate a capability.
+
+Luckily Justin, who actually did the hard work of making my proof of concept code production read, uses this in `smart_proxy_pulp`. The use case for Pulp 3 was various file types. In Katello with Pulp 2 there's a static configuration in `katello.yaml`, but with Pulp 3 we can query the Pulp API for supported file types and communicate this. Less configuration and more autodetection is a good user experience.
+
+Implementation is straight forward using the `capability` in the Proxy Plugin DSL. The static example is not actually in the code, but it illustrates the initial design.
+
+```ruby
+module PulpProxy
+  class Pulp3Plugin < ::Proxy::Plugin
+    plugin "pulp3", ::PulpProxy::VERSION
+
+    # ...
+
+    # Static capability
+    capability :pulp3
+
+    # Dynamic capability
+    capability( lambda do ||
+      Pulp3Client.capabilities
+    end)
+  end
+end
+```
+
+## Updating Foreman
+
+The Foreman registration also needed to be updated. Previously it stored a Smart Proxy entity with n Feature entities. To deal with the new datastructures the Feature entity was expanded to store a JSON serialized array of capabilities and a JSON serialized hash of settings. When refreshing features it first tries the `/v2/features` endpoint. Older proxies will return a 404 and the older `/features` endpoint is queried. In this case no settings nor capabilities are available.
```

Comment:
s/with n Feature/with Feature/

Reply:
do we also store/cache status somehow? and do we keep timestamp for capabilities and settings?

Reply:
AFAIK we don't, but it does make sense. Currently we store an empty hash, but when it's loaded from v1, but that could be nil so you can detect it. That could prevent downgrades.

Reply:
probably the information about cache should be mentioned here so people know they shouldn't abuse the call